Documentation/capabilities-guide.md

```diff
@@ -319,6 +320,94 @@ image: using image from local store for image name quay.io/coreos/alpine-sh
 listening on [::]:80 ...
 ```
 
+## Capabilities when running as non-root
+
+The capability isolators (and default capabilities) mentioned in this document operate on the capability bounding set.
+When running containers as non-root, capabilities are not added to the effective set of the process, which is the one the kernel checks when doing privileged operations.
+This means the process won't be able to run the privileged operations enabled by the capabilities granted to the container directly.
+
+For example, including `CAP_NET_RAW` in the retain set when running a container as non-root doesn't allow the container to run `ping` (which uses raw sockets):
+
+```
+$ sudo rkt run --interactive kinvolk.io/aci/busybox --user=1000 --group=1000 --caps-retain=CAP_NET_RAW --exec ping -- 8.8.8.8
+PING 8.8.8.8 (8.8.8.8): 56 data bytes
+ping: permission denied (are you root?)
+```
+
+To be able to execute `ping` as a non-root user, the binary needs to have the corresponding file capability.
+
+Also, running an image with file capabilities requires disabling seccomp in rkt.
+This is due to a systemd bug where using seccomp results in enabling [no_new_privs][NNP] (you can track progress in [systemd/systemd#6763](https://github.com/systemd/systemd/pull/6763)).
+
+### Building images with file capabilities
+
+Building images that include files with file capabilities is challenging since:
+
+* [build][acbuild] doesn't preserve file capabilities. See [containers/build#197](https://github.com/containers/build/issues/197).
+* docker build doesn't preserve file capabilities. See [moby/moby#35699](https://github.com/moby/moby/issues/35699).
+
+However, provided we have an ACI (created for example with build or with docker2aci), we can extract it and add the file capabilities manually.
+
+#### Example
+
+We'll use build to create an Ubuntu ACI with ping installed:
+
+```bash
+#!/usr/bin/env bash
+
+acbuild --debug begin docker://ubuntu
+acbuild --debug set-name example.com/filecap
+
+# Install ping
+acbuild --debug run -- apt-get update
+acbuild --debug run -- apt-get install -y inetutils-ping
+
+# ping comes as a setuid file, we don't want that
+acbuild --debug run -- chmod -s /bin/ping
+
+acbuild --debug set-exec /bin/bash
+
+acbuild --debug write --overwrite filecaps.aci
+```
+
+After running the script, we'll extract the ACI, add file capabilities, and rebuild it.
+We need to use [bsdtar][bsdtar] since GNU tar doesn't support extended attributes:
```

Comment:
Isn't this covered by `tar --xattrs`? I seem to remember having used it in the past.

Reply:
It seems `--xattrs` at compression time and `--xattrs --xattrs-include='*.*'` at decompression time does the trick.